### PR TITLE
Simpler checked shifts in MIR building

### PR DESCRIPTION
**Pull request by scottmcm**
Doing masking to check unsigned shift amounts is overcomplicated; just comparing the shift directly saves a statement and a temporary, as well as is much easier to read as a human.  And shifting by unsigned is the canonical case -- notably, all the library shifting methods (that don't support every type) take shift RHSs as `u32` -- so we might as well make that simpler since it's easy to do so.

This PR also changes *signed* shift amounts to `IntToInt` casts and then uses the same check as for unsigned.  The bit-masking is a nice trick, but for example LLVM actually canonicalizes it to an unsigned comparison anyway <https://rust.godbolt.org/z/8h59fMGT4> so I don't think it's worth the effort and the extra `Constant`.  (If MIR's `assert` was `assert_nz` then the masking might make sense, but when the `!=` uses another statement I think the comparison is better.)

To review, I suggest looking at https://github.com/rust-lang/rust/pull/109475/commits/2ee0468c49be9f8ea68b7b935ea1f4c2555849e9 first -- that's the interesting code change and has a MIR diff.

My favourite part of the diff:
```diff
-        _20 = BitAnd(_19, const 340282366920938463463374607431768211448_u128); // scope 0 at $DIR/shifts.rs:+2:34: +2:44
-        _21 = Ne(move _20, const 0_u128); // scope 0 at $DIR/shifts.rs:+2:34: +2:44
-        assert(!move _21, "attempt to shift right by `{}`, which would overflow", _19) -> [success: bb3, unwind: bb7]; // scope 0 at $DIR/shifts.rs:+2:34: +2:44
+        _18 = Lt(_17, const 8_u128);     // scope 0 at $DIR/shifts.rs:+2:34: +2:44
+        assert(move _18, "attempt to shift right by `{}`, which would overflow", _17) -> [success: bb3, unwind: bb7]; // scope 0 at $DIR/shifts.rs:+2:34: +2:44
```

### CODE REVIEW

File: tests/mir-opt/building/shifts.shift_signed.built.after.mir [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    bb0: {
+        StorageLive(_6);                 // scope 0 at $DIR/shifts.rs:+2:9: +2:45
+        StorageLive(_7);                 // scope 0 at $DIR/shifts.rs:+2:10: +2:20
+        StorageLive(_8);                 // scope 0 at $DIR/shifts.rs:+2:10: +2:15
+        _8 = _1;                         // scope 0 at $DIR/shifts.rs:+2:10: +2:15
+        StorageLive(_9);                 // scope 0 at $DIR/shifts.rs:+2:19: +2:20
+        _9 = _3;                         // scope 0 at $DIR/shifts.rs:+2:19: +2:20
+        _10 = _9 as u8 (IntToInt);       // scope 0 at $DIR/shifts.rs:+2:10: +2:20
+        _11 = Ge(move _10, const 8_u8);  // scope 0 at $DIR/shifts.rs:+2:10: +2:20
+        assert(!move _11, "attempt to shift right by `{}`, which would overflow", _9) -> [success: bb1, unwind: bb7]; // scope 0 at $DIR/shifts.rs:+2:10: +2:20
```
Comment (WaffleLapkin) [inline tests/mir-opt/building/shifts.shift_signed.built.after.mir:52]: Would it make more sense to use `Lt` and `expected = true`, such that this becomes `_11 = Lt...; assert(move _11, ...)`?

**Review by WaffleLapkin — COMMENTED** [review]
This looks good to me! Can you please squash 168f713 and 1cb37c2?

File: tests/mir-opt/building/shifts.shift_signed.built.after.mir (hunk 1 above)
Comment (scottmcm) [reply tests/mir-opt/building/shifts.shift_signed.built.after.mir]: Oh, nice, that's much more readable.  (I hadn't even noticed that `bool` 😬)

Updated and squashed.

File: compiler/rustc_mir_build/src/build/expr/as_rvalue.rs [hunk 2]
```diff
@@ -566,41 +566,51 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
 
-                let outer_bits = self.temp(rhs_ty, span);
-                self.cfg.push_assign(
-                    block,
-                    source_info,
-                    outer_bits,
-                    Rvalue::BinaryOp(BinOp::BitAnd, Box::new((rhs.to_copy(), mask))),
-                );
-
                 let overflows = self.temp(bool_ty, span);
```
[the remark this answers is not in the document]
Comment (scottmcm) [reply compiler/rustc_mir_build/src/build/expr/as_rvalue.rs]: So it is!

**Review by WaffleLapkin — APPROVED** [review]
I should be sleeping and not approving prs rn but here we are

**Comment by WaffleLapkin** [comment]
@bors r+ rollup
(the queue is long, so trying to roll everything up)

**Pull request merged**